### PR TITLE
[SDK-38] fixes deltas on browsers in ably-js

### PR DESCRIPTION
**Pull request by SimonWoolf**
And fixes a few of other issues issues I noticed while doing it.

Also implements what we decided in https://github.com/ably/docs/issues/891 for what if you have no vcdiff plugin (or are using a browser too old to support deltas).

Related error code pr: https://github.com/ably/ably-common/pull/56

### CODE REVIEW

File: spec/realtime/delta.test.js [hunk 1]
```diff
@@ -45,23 +45,33 @@ define(['shared_helper', 'vcdiff-decoder'], function(helper, vcdiffDecoder) {
...
-					test.equal(testVcdiffDecoder.numberOfCalls, testData.length - 1, 'Check number of delta messages');
+			channel.attach(function(err) {
+				if(err) {
+					test.ok(false, displayError(err));
 					closeAndFinish(test, realtime);
 				}
-			});
 
-			realtime.connection.on('connected', function() {
+				channel.on('attaching', function(stateChange) {
```
Comment (paddybyers) [inline spec/realtime/delta.test.js:61]: Wouldn't it be more straightforward just to test that you actually receive deltas (eg by checking extras) ?

Edit: I see you did that later

File: common/lib/client/realtimechannel.js [hunk 2]
```diff
@@ -83,7 +86,24 @@ var RealtimeChannel = (function() {
 		if (this._decodingContext)
 			this._decodingContext.channelOptions = this.channelOptions;
 		if(this._shouldReattachToSetOptions(options)) {
-			this._attach(true, null, callback);
+			/* This does not just do _attach(true, null, callback) because that would put us
```
Comment (paddybyers) [inline common/lib/client/realtimechannel.js:89]: Does this need a spec clarification?

Comment (SimonWoolf) [reply common/lib/client/realtimechannel.js:89]: yeah, we should

File: common/lib/types/message.js [hunk 3]
```diff
@@ -187,22 +187,30 @@ var Message = (function() {
...
+							try {
+								var deltaBase = context.baseEncodedPreviousPayload;
+								if(typeof deltaBase === 'string') {
+									deltaBase = BufferUtils.utf8Encode(deltaBase);
 								}
-								continue;
-							} else {
-								throw new Error('Missing Vcdiff decoder (https://github.com/ably-forks/vcdiff-decoder)');
+
+								/* vcdiff expects Uint8Arrays, can't copy with ArrayBuffers. (also, if we
```
Comment (QuintinWillison) [inline common/lib/types/message.js:202]: This provided me food for thought. Nothing to stop us merging now but, FWIW: https://github.com/ably/ably-js/issues/659

File: common/lib/client/realtimechannel.js (hunk 2 above)
Comment (QuintinWillison) [reply common/lib/client/realtimechannel.js:89]: I've created https://github.com/ably/docs/issues/894 

@SimonWoolf, my only concern on a skim read of this code (sorry, I've not looked in great depth) is to ensure that you've considered that the `_shouldReattachToSetOptions` 'method' implementation returns `true` for both the `attaching` as well as the `attached` local states, per [RTL16a](https://docs.ably.io/client-lib-development-guide/features/#RTL16a). Would this code running when the channel is `attaching` have a different outcome?

Comment (QuintinWillison) [reply common/lib/client/realtimechannel.js:89]: [moved](https://ably-real-time.slack.com/archives/C030C5YLY/p1591598178340200) to internal Slack channel

**Pull request merged**